### PR TITLE
Improve algorithm

### PR DESCRIPTION
**Pull request by Atlasoin**
- src/*: modify algorithm to support fft1d, fft2d and dct
- test/*: a simple test file for stego.js

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,6 +23,8 @@
     "clean": "rm -rf umd es"
   },
   "dependencies": {
+    "@rollup-umd/build": "^1.1.1",
```
Comment (guanbinrui) [inline package.json]: Move `@rollup-umd/build` to `devDependencies`?

File: package.json [hunk 2]
```diff
@@ -23,6 +23,8 @@
...
+    "canvas": "^2.6.1",
```
Comment (guanbinrui) [inline package.json]: This lib only used in CLI shouldn't be listed here. CI will install it all the time.

Comment (Atlasoin) [reply package.json]: But if a user download the repo and want to run it locally, only the commands `npm install` and `npm run build` will fail without canvas lib in dependencies.

**Comment by septs** [comment]
@AtlasChuen plz sync master branch.
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

File: package.json (hunk 2 above)
Comment (septs) [reply package.json]: see https://docs.npmjs.com/files/package.json#optionaldependencies

**Comment by septs** [comment]
Please maintain backward compatibility at the design level.
Breaking backward compatibility will result in **Maskbook** not being able to decode past content

or

original logic move to `src/v1` folder
new design encoder and decoder in `src/v2` folder

File: package.json (hunk 2 above)
Comment (guanbinrui) [reply package.json]: Already moved to devDeps in the `master` branch.

File: src/flag.ts [hunk 3]
```diff
@@ -25,13 +31,30 @@ export interface Flags {
...
+        t = DEFAULT_FFT2D_TOLERANCE;
+        break;
+      case TransformAlgorithm.DCT:
+        t = DEFAULT_DCT_TOLERANCE;
+        break;
+      default:
+        break;
+    }
+  else
+    t = parseInt(tolerance, 10);
```
Comment (septs) [inline src/flag.ts]:
```suggestion
  const transform2tolerance = {
    [TransformAlgorithm.FFT1D]: DEFAULT_FFT1D_TOLERANCE,
    [TransformAlgorithm.FFT2D]: DEFAULT_FFT2D_TOLERANCE,
    [TransformAlgorithm.DCT]: DEFAULT_DCT_TOLERANCE,
  }
  const tolerance = flags.tolerance === TOLERANCE_NOT_SET
    ? transform2tolerance[flags.transform]
    : parseInt(flags.tolerance, 10);
```

**Review by septs — CHANGES_REQUESTED** [review]

File: scripts/test-stego.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
+#This is a script to test stegojs locally
```
Comment (septs) [inline scripts/test-stego.sh:1]: plz using https://mochajs.org or https://jestjs.io unit test framework

mocha using reference:
https://github.com/DimensionDev/common-protocols

jest using reference:
https://github.com/DimensionDev/coda

**Review by septs — CHANGES_REQUESTED** [review]

**Comment by Atlasoin** [comment]
I randomly tested a small number of pictures to encode and upload to FB and decode. It seems generally with selecting appropriate image, current v2 algorithm can resist the compression by Facebook. But I noticed some images can be resized by FB for some unknown reason (possibly because of the high resolution of the image), in which case decode will fail.

Btw, what's the reason for CI test fail?

**Comment by septs** [comment]
@AtlasQuan request rebase master.

**Comment by septs** [comment]
i use #10 merge to master.

**Pull request closed without merging**